### PR TITLE
Answer:7 Power of effects

### PR DESCRIPTION
**Pull request by LMFinney**
I'm not sure if I achieved step 4 - I'm not sure what you intended.
We could have injected the PushService directly wherever the injection token was used, right?

## Checklist for challenge submission

- [x] Start your PR message with Answer:${challenge_number}

## Warning:

- If you want feedback or review, you must support the project on GitHub:
  - [$5 per review](https://github.com/sponsors/tomalaforge)
  - [$25 for lifetime reviews](https://github.com/sponsors/tomalaforge)
  - Create one of more challenges
  - Have many valuable contributions to the project

Alternatively, you can still submit your PR to join the list of answered challenges or to be reviewed by a community member. 🔥

### CODE REVIEW

**Comment by jdegand** [comment]
You don't have to worry about lazy loading because you used the lifecycle methods of component store.  
In the initial app, all the lists are constantly growing.

The injection token approach makes more sense if you don't (or can't) convert all components to use Component Stores.

File: apps/ngrx/7-power-of-effect/src/app/app.effects.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  displayAlerts$ = createEffect(
+    () =>
+      this.actions$.pipe(
+        ofType(
+          schoolActions.addOneSchool,
+          studentActions.addOneStudent,
+          teacherActions.addOneTeacher,
+        ),
+        map((action) => {
+          switch (action.type) {
```
Comment (tomalaforge) [inline apps/ngrx/7-power-of-effect/src/app/app.effects.ts:28]: I'm not a big fan of switch statement in effect. You should create separate effects for each use cases

File: apps/ngrx/7-power-of-effect/src/app/push-token.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+import { Push } from '@angular-challenges/power-of-effect/model';
+import { inject, InjectionToken } from '@angular/core';
+import { Observable } from 'rxjs';
+
+export const PUSH_TOKEN = new InjectionToken<Observable<Push | undefined>>(
+  'PUSH_TOKEN',
+  {
+    factory: () => inject(PushService).notification$,
+  },
+);
```
Comment (tomalaforge) [inline apps/ngrx/7-power-of-effect/src/app/push-token.ts:11]: That's the tricky part and you did it well. Nice job.

File: apps/ngrx/7-power-of-effect/src/app/school/school.store.ts [hunk 3]
```diff
@@ -45,4 +47,20 @@ export class SchoolStore
   ngrxOnStoreInit() {
     this.loadSchools();
   }
+
+  ngrxOnStateInit() {
+    this.effect(() =>
+      inject(PUSH_TOKEN).pipe(
```
Comment (tomalaforge) [inline apps/ngrx/7-power-of-effect/src/app/school/school.store.ts:53]: That's the second point I want to highly with this challenge, it's that we can use any observable inside an effect and not only the Ngrx action observbale. Effect is just a subscription. 

Nice work

**Comment by tomalaforge** [comment]
Very nice work and I think you are the first one to complete the challenge. I was going to suppress it 😅

File: apps/ngrx/7-power-of-effect/src/app/app.effects.ts (hunk 1 above)
Comment (LMFinney) [reply apps/ngrx/7-power-of-effect/src/app/app.effects.ts]: I just pushed two commits:

1. One to split the effect into four effects (1 for each of the three notification types, and one to do the actual alert).
2. Converted to functional effects (just for practice).

File: apps/ngrx/7-power-of-effect/src/app/app.effects.ts [hunk 4]
```diff
@@ -24,10 +20,12 @@ export class AppEffects {
       }),
       map((message) => appApiActions.alert({ message })),
     ),
-  );
+  { functional: true },
+);
```
Comment (tomalaforge) [inline apps/ngrx/7-power-of-effect/src/app/app.effects.ts:24]: 👍

**Pull request closed without merging**